### PR TITLE
Feature/amqp reconnect

### PR DESCRIPTION
**Pull request by Ashhm**
Reconnection mechanism implementation

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,21 +1,122 @@
...
- * @param {Object} options Connection socket options
+ * @param {Object} socketOptions Connection socket options
  * @returns {*}
  */
-module.exports = function createService(url, options) {
+module.exports = function createService(url, socketOptions) {
   return {
     name: 'amqp',
+    metadata: {
+      reconnectTimeout: 5000,
```
Comment (ghost) [inline src/index.js:23]: Maybe extract to `settings` ? Also consider ENV variable? WDYT?

File: src/index.js [hunk 2]
```diff
@@ -1,21 +1,122 @@
...
+      /**
+       * Connect to rabbitMQ with socketOptions.
+       * Reconnect is supported
+       *
+       * @returns {Promise<void>}
+       * @private
+       */
+      async _connect() {
+        let connectionTimeout;
+        const reconnect = () => {
```
Comment (ghost) [inline src/index.js]: Maybe extract this to methods?

File: src/index.js [hunk 3]
```diff
@@ -34,6 +135,7 @@ module.exports = function createService(url, options) {
         return this.channel
```
Comment (ghost) [inline src/index.js:132]: Maybe we need to check availability of rabbit and channel somehow, to ensure we are not in "reconnect" mode. WDYT?

Comment (ghost) [reply src/index.js:132]: And the same for sendToQueue

Comment (Ashhm) [reply src/index.js:132]: I thought about that, probably we should have some mechanism to re-send all of these messages which are failed to send when reconnection is done

File: src/index.js (hunk 1 above)
Comment (Ashhm) [reply src/index.js:23]: As I know, moleculer does not allow us to change this value in amqp transport (or disable reconnection at all), so I think we should not allow this also.

File: src/index.js (hunk 3 above)
Comment (ghost) [reply src/index.js:132]: Maybe, it can be done separately then.

File: src/index.js (hunk 1 above)
Comment (ghost) [reply src/index.js:23]: Let's follow their practice then )

File: src/index.js (hunk 2 above)
Comment (Ashhm) [reply src/index.js]: Done, check pls https://github.com/Ashhm/moleculer-amqp/pull/7/commits/319de97836594a87319cbc59729ad8861f072141

File: src/index.js (hunk 3 above)
Comment (Ashhm) [reply src/index.js:132]: Yea, need to brainstorm about how to deliver delayed response

File: src/index.js [hunk 4]
```diff
@@ -1,21 +1,119 @@
...
+        }
+      },
+
+      /**
+       * Reconnect to rabbitMQ with timeout
+       *
+       * @private
+       */
+      _reconnect() {
+        if (!this._isReconnecting) {
```
Comment (ghost) [inline src/index.js:68]: Looks like better approach !

File: test/index.spec.js [hunk 5]
```diff
@@ -178,4 +185,227 @@ describe('AMQP queues', () => {
...
+          && expect(service._connect).to.have.been.calledOnce;
+      });
+
+      it('should call amqplib connect',
+        () => expect(amqplib.connect).to.have.been.called);
+
+      it('should set service channel', () => expect(service.channel).to.exist);
+    });
+
+    describe('on multiply errors', () => {
```
Comment (ghost) [inline test/index.spec.js]: `on multiplE errors` is more correct :)

**Review by ghost — APPROVED** [review]
Awesome! Thank you

File: test/index.spec.js (hunk 5 above)
Comment (Ashhm) [reply test/index.spec.js]: Fixed

**Pull request merged**